managed/services/uievents/uievents.go

```diff
@@ -0,0 +1,262 @@
+}
+
+func (s *Service) processComponentsStat(metrics []*pmmv1.ServerMetric_Metric) []*pmmv1.ServerMetric_Metric {
+	type Stat struct {
+		SlowComponentsP95_1s  []string `json:"slow_components_p95_1s"`
+		SlowComponentsP95_5s  []string `json:"slow_components_p95_5s"`
+		SlowComponentsP95_10s []string `json:"slow_components_p95_10s"`
+	}
+
+	if len(s.componentsUsage) > 0 {
+		componentsStat := &Stat{
+			SlowComponentsP95_1s:  []string{},
+			SlowComponentsP95_5s:  []string{},
+			SlowComponentsP95_10s: []string{},
+		}
+
+		// SlowComponentP95
+		keys := make([]string, 0, len(s.componentsUsage))
+		for key := range s.componentsUsage {
+			keys = append(keys, key)
+		}
+		sort.SliceStable(keys, func(i, j int) bool {
+			return s.componentsUsage[keys[i]].loadTime.ValueAtPercentile(95) > s.componentsUsage[keys[j]].loadTime.ValueAtPercentile(95)
+		})
+		for i := 0; i < len(keys); i++ {
+			sortedKey := keys[i]
+			stat := s.componentsUsage[sortedKey]
+			p95 := stat.loadTime.ValueAtPercentile(95)
+			if p95 >= 1_000 {
+				componentsStat.SlowComponentsP95_1s = append(componentsStat.SlowComponentsP95_1s, stat.uid)
+			}
+			if p95 >= 5_000 {
+				componentsStat.SlowComponentsP95_5s = append(componentsStat.SlowComponentsP95_5s, stat.uid)
+			}
+			if p95 >= 10_000 {
+				componentsStat.SlowComponentsP95_10s = append(componentsStat.SlowComponentsP95_10s, stat.uid)
+			}
+		}
+
+		marshal, err := json.Marshal(componentsStat)
+		if err != nil {
+			s.l.Error("failed to marshal to JSON")
+		}
+		metrics = append(metrics, &pmmv1.ServerMetric_Metric{
+			Key:   "ui-events-components",
+			Value: string(marshal),
+		})
+	}
+	return metrics
+}
+
+// Store stores metrics for further processing and sending to Portal.
+func (s *Service) Store(ctx context.Context, request *uievents.StoreRequest) (*uievents.StoreResponse, error) {
+	for _, event := range request.DashboardUsage {
+		stat, ok := s.dashboardUsage[event.Uid]
+		if !ok {
+			stat = &DashboardUsageStat{
+				title:    event.Title,
+				uid:      event.Uid,
+				useCount: 0,
+				loadTime: s.loadTimeHistogram(),
+			}
+			s.dashboardUsage[event.Uid] = stat
+		}
+		stat.useCount = stat.useCount + 1
+		err := stat.loadTime.RecordValue(int64(event.LoadTime))
+		if err != nil {
+			s.l.Error("failed to record value", err)
+		}
+	}
+	for _, event := range request.Fetching {
+		stat, ok := s.componentsUsage[event.Component]
+		if !ok {
+			stat = &ComponentsUsageStat{
+				uid:      event.Component,
+				useCount: 0,
+				loadTime: s.loadTimeHistogram(),
+			}
+			s.componentsUsage[event.Component] = stat
+		}
+		stat.useCount = stat.useCount + 1
+		err := stat.loadTime.RecordValue(int64(event.LoadTime))
+		if err != nil {
+			s.l.Error("failed to record value", err)
+		}
+	}
+
+	return &uievents.StoreResponse{}, nil
+}
+
+func (s *Service) loadTimeHistogram() *hdrhistogram.Histogram {
+	lowersValue := int64(0)
+	highestValue := int64(1_000 * 60 * 10) // 10 min
+	numberOfSignificantValueDigits := 5
```

Comment:
seems like these values can be constants

Reply:
in that case using `loadTimeHistogram` can be replaced by calling `hdrhistogram.New(lowersValue, highestValue, numberOfSignificantValueDigits)`

Reply:
it just a factory method, and good practice, we should try to minimize scope of variables/constants. that is why extracting makes a little sense.